R/selectors.R

```diff
@@ -519,27 +523,22 @@ cs__date <- function() {
 #'
 #' # Select all columns except for datetime columns:
 #' df$select(!cs$datetime())
-cs__datetime <- function(time_unit = c("ms", "us", "ns"), time_zone) {
+cs__datetime <- function(time_unit = c("ms", "us", "ns"), time_zone = list("*", NULL)) {
   time_unit <- arg_match(time_unit, values = c("ms", "us", "ns"), multiple = TRUE)
-  datetime_dtypes <- list()
-  # When time_zone is missing, we want to detect all datetime, no matter the
-  # presence/absence or the value of the timezone.
-  if (missing(time_zone)) {
-    for (tu in time_unit) {
-      datetime_dtypes <- append(datetime_dtypes, pl$Datetime(time_unit = tu, time_zone = "*"))
-      datetime_dtypes <- append(datetime_dtypes, pl$Datetime(time_unit = tu, time_zone = NULL))
-    }
-  } else if (is.null(time_zone)) {
-    for (tu in time_unit) {
-      datetime_dtypes <- append(datetime_dtypes, pl$Datetime(time_unit = tu, time_zone = NULL))
-    }
-  } else {
-    for (tz in time_zone) {
-      for (tu in time_unit) {
-        datetime_dtypes <- append(datetime_dtypes, pl$Datetime(time_unit = tu, time_zone = tz))
-      }
-    }
+  if (!is_character(time_zone) && is_list(time_zone) && is.null(time_zone)) {
+    abort("`time_zone` must be a character vector, `NULL`, or a list.")
   }
+  if (is.null(time_zone)) {
+    time_zone <- list(NULL)
+  }
+
+  datetime_dtypes <- time_unit |>
+    lapply(\(tu) {
+      time_zone |>
+        lapply(\(tz) pl$Datetime(tu, tz))
```

Comment:
The error when an invalid time zone is entered here is fairly difficult to understand, so there may be future improvement potential.

Reply:
FYI this doesn't error in py-polars:

```
import polars as pl
import polars.selectors as cs
from datetime import datetime
from zoneinfo import ZoneInfo

tokyo_tz = ZoneInfo("Asia/Tokyo")

df = pl.DataFrame(
    {
        "tstamp_tokyo": [
            datetime(1999, 7, 21, 5, 20, 16, 987654, tzinfo=tokyo_tz),
            datetime(2000, 5, 16, 6, 21, 21, 123465, tzinfo=tokyo_tz),
        ]
    }
)

df.select(cs.datetime(time_zone = "foobar"))
```
Or did you mean for instance input like `time_zone = 1`?

Reply:
> Or did you mean for instance input like time_zone = 1?

Ah no that would be caught earlier

Reply:
Oh, sorry for the lack of clarity. It means that if we pass a list as an argument, each element of the list must be a character of length 1.

```r
# error
cs$datetime(time_zone = list(c("UTC", "Asia/Tokyo")))
# ok
cs$datetime(time_zone = c("UTC", "Asia/Tokyo"))
```